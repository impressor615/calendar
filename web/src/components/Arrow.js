import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';


const Arrow = ({ direction, onClick, ...rest }) => (
  <button
    {...rest}
    onClick={onClick}
    className={classnames(
      'arrow',
      rest.className,
      { [direction]: direction },
    )}
  />
);

Arrow.defaultProps = {
  direction: 'previous',
};

Arrow.propTypes = {
  direction: PropTypes.oneOf(['next', 'previous']),
  onClick: PropTypes.func.isRequired,
};

export default Arrow;