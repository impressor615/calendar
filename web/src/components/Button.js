import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';


const Button = ({ children, color, onClick, active, ...rest }) => (
  <button
    type="button"
    onClick={onClick}
    {...rest}
    className={classnames(
      'btn',
      rest.className,
      { [color]: color },
      { active },
    )}>
    { children }
  </button>
);

Button.defaultProps = {
  color: '',
  active: false,
  onClick: () => {},
};

Button.propTypes = {
  color: PropTypes.oneOf(['', 'primary', 'secondary']),
  onClick: PropTypes.func,
  children: PropTypes.node.isRequired,
  active: PropTypes.bool,
}

export default Button;
