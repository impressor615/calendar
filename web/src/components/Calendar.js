import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import { eventBind } from 'utils/eventUtils';
import CONSTANTS from '../constants';


export const EventBlock = ({ children, onToggle }) => (
  <button
    type="button"
    className="event-block btn"
    onClick={onToggle}
  >
    {children}
  </button>
);

const { DAY, HOURS } = CONSTANTS;
export const CalendarDay = ({
  day,
  dateObj,
  children,
  currentDate,
  onToggle,
}) => {
  const isYearShowing = currentDate.year() !== dateObj.year;
  const isMonthShowing = (currentDate.month() + 1) !== dateObj.month;
  return (
    <div
      className="cal-block"
      onClick={eventBind(onToggle, { type: 'month', dateObj })}
    >
      { day ? <div className="day">{day}</div> : null}
      <div>
        { isYearShowing ? `${dateObj.year}년 ${dateObj.month}월 ${dateObj.date}일` : null }
        { !isYearShowing && isMonthShowing ? `${dateObj.month}월 ${dateObj.date}일` : null }
        { !isYearShowing && !isMonthShowing ? `${dateObj.date}` : null }
      </div>
      {children}
    </div>
  );
}

export const MonthCalendar = ({ range, onToggle, currentDate, data }) => (
  <section className="month-calendar">
    {
      range.map((dateObj, index) => {
        const dataKey = dateObj.moment.format('YYYY-MM-DD');
        const events = data[dataKey];
        return (
          <CalendarDay
            key={index.toString()}
            currentDate={currentDate}
            dateObj={dateObj}
            day={DAY[index]}
            onToggle={onToggle}
          >
           {
              events
                ? (
                  <Fragment>
                    {
                      Object.entries(events).map(([key, event], index) => (
                        <EventBlock
                          key={index.toString()}
                          onToggle={eventBind(onToggle, { type: 'event', dateObj: event })}
                        >
                          &#9675; {`${key}:00 ${event.title}`}
                        </EventBlock>
                      ))
                    }
                  </Fragment>
                ) : null
            }
          </CalendarDay>
        );
      })
    }
  </section>
);

export const CalendarTime = ({ children, onToggle }) => (
  <button className="cal-block sm" onClick={onToggle}>
    {children}
  </button>
);

export const WeekCalendar = ({ range, onToggle, currentDate }) => (
  <section className="week-calendar">
    <div className="cal-block sm" />
    {
      range.map((dateObj, index) => {
        const isYearShowing = currentDate.year() !== dateObj.year;
        const isMonthShowing = (currentDate.month() + 1) !== dateObj.month;
        return (
          <div className="cal-block sm" key={index.toString()}>
            <div className="day">{DAY[index]}</div>
            <div>
              { isYearShowing ? `${dateObj.year}년 ${dateObj.month}월 ${dateObj.date}일` : null }
              { !isYearShowing && isMonthShowing ? `${dateObj.month}월 ${dateObj.date}일` : null }
              { !isYearShowing && !isMonthShowing ? `${dateObj.date}` : null }
            </div>
          </div>
        );
      })
    }
    {
      HOURS.map((hourObj, index) => (
        <Fragment key={index.toString()}>
          <div className="cal-block sm">
            <span className="hour">{hourObj.text}</span>
          </div>
          {
            range.map((dateObj, index) => (
              <CalendarTime
                key={index.toString()}
                onToggle={eventBind(onToggle, { type: 'week', hour: hourObj.code, dateObj })}
              />
            ))
          }
        </Fragment>
      ))
    }
  </section>
);

const Calendar = ({
  type,
  range,
  currentDate,
  onToggle,
  data,
}) => (
  <article className="calendar">
    {
      type === 'month'
        ? <MonthCalendar range={range} onToggle={onToggle} currentDate={currentDate} data={data} />
        : <WeekCalendar range={range} onToggle={onToggle} currentDate={currentDate} data={data} />
    }
  </article>
);

Calendar.propTypes = {
  type: PropTypes.string.isRequired,
  currentDate: PropTypes.object.isRequired,
  range: PropTypes.arrayOf(PropTypes.object).isRequired,
  data: PropTypes.object.isRequired,
  onToggle: PropTypes.func.isRequired,
};

export default Calendar;
