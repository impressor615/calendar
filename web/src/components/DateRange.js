import React from 'react';
import PropTypes from 'prop-types';
import DatePicker from 'react-datepicker';

const DateRange = ({ startDate, endDate, onChange }) => (
  <div className="date-range">
    <DatePicker
      showTimeSelect
      selected={startDate}
      onChange={onChange}
      timeIntervals={60}
      timeCaption="time"
      dateFormat="YYYY-M-D h:mm A"
    />
    <div>~</div>
    <div>{ endDate.format('YYYY-M-D h:mm A') }</div>
  </div>
);

DateRange.propTypes = {
  startDate: PropTypes.object.isRequired,
  endDate: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired,
}

export default DateRange;
