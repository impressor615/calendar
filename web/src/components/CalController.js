import React from 'react';
import PropTypes from 'prop-types';

import { withContext } from 'AppContext';
import Arrow from 'components/Arrow';
import Button from 'components/Button';
import { getCalendarTitle } from 'utils/dateUtils';


const CalController = ({
  type,
  currentDate,
  onArrowClick,
  onTypeChange,
}) => (
  <article className="controller">
    <div className="controller-wrapper">
      <section className="date-control">
        <Arrow name="previous" onClick={onArrowClick} />
        <div className="title">{getCalendarTitle(currentDate, type)}</div>
        <Arrow name="next" direction="next" onClick={onArrowClick} />
      </section>
      <section className="type-control">
        <Button name="month" onClick={onTypeChange} color="primary" active={type === 'month'}>월</Button>
        <Button name="week" onClick={onTypeChange} color="primary" active={type === 'week'}>주</Button>
      </section>
    </div>
  </article>
);

CalController.propTypes = {
  type: PropTypes.string.isRequired,
  currentDate: PropTypes.object.isRequired,
  onArrowClick: PropTypes.func.isRequired,
  onTypeChange: PropTypes.func.isRequired,
}

export default withContext(CalController);
export const PureCalController = CalController;
