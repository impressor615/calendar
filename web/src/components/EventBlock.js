import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Button from 'components/Button';


const EventBlock = ({
  children,
  onToggle,
  onDragStart,
  fluid,
  ...rest
}) => (
  <Button
    {...rest}
    draggable
    onClick={onToggle}
    onDragStart={onDragStart}
    className={classnames(
      'event-block',
      rest.className,
      { fluid },
    )}
  >
    { children }
  </Button>
);

EventBlock.defaultProps = {
  fluid: false,
};

EventBlock.propTypes = {
  children: PropTypes.node.isRequired,
  onToggle: PropTypes.func.isRequired,
  onDragStart: PropTypes.func.isRequired,
  fluid: PropTypes.bool,
}

export default EventBlock;