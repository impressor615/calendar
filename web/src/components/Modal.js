import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Button from 'components/Button';

export const ModalHeader = ({ children, onToggle }) => (
  <div className="modal-header">
    <div>{children}</div>
    <Button onClick={onToggle}>&times;</Button>
  </div>
);

ModalHeader.propTypes = {
  children: PropTypes.node.isRequired,
  onToggle: PropTypes.func.isRequired,
};

export const ModalBody = ({ children }) => (
  <div className="modal-body">
    {children}
  </div>
);

ModalBody.propTypes = {
  children: PropTypes.node.isRequired,
};

export const ModalFooter = ({ children }) => (
  <div className="modal-footer">
    { children }
  </div>
);

ModalFooter.propTypes = {
  children: PropTypes.node.isRequired,
};

export const ModalDialog = ({ children, ...rest }) => (
  <div className={classnames('modal-dialog', rest.className)}>
    { children }
  </div>
);

ModalDialog.propTypes = {
  children: PropTypes.node.isRequired,
};

const Modal = ({ isOpen, children, backdrop, onToggle, ...rest }) => (
  isOpen ? (
    ReactDOM.createPortal(
      <div
        {...rest}
        className={classnames('modal', rest.className)}
      >
        {children}
        { backdrop ? <div className="modal-backdrop" onClick={onToggle} /> : null }
      </div>,
      document.getElementById('modal-root'),
    )
  ) : null
);

Modal.defaultProps = {
  backdrop: true,
  onToggle: () => {},
};

Modal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired,
  onToggle: PropTypes.func,
  backdrop: PropTypes.bool,
};

export default Modal;
