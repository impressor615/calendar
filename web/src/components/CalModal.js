import React from 'react';
import PropTypes from 'prop-types';

import Modal, { ModalHeader, ModalBody, ModalFooter } from 'components/Modal';
import DateRange from 'components/DateRange';


const CalModal = ({
  isOpen,
  onToggle,
  onChange,
  onDateChange,
  onSubmit,
  event,
}) => (
  <Modal
    isOpen={isOpen}
    className="cal-modal"
  >
    <ModalHeader onToggle={onToggle}>
      일정 추가 및 변경
    </ModalHeader>
    <ModalBody>
      <input
        id="title"
        type="text"
        placeholder="일정 제목을 입력해주세요!"
        onChange={onChange}
        value={event.title}
      />
      <DateRange
        startDate={event.start_date}
        endDate={event.end_date}
        onChange={onDateChange}
      />
    </ModalBody>
    <ModalFooter>
      <button
        className="btn primary"
        onClick={onToggle}
      >
        취소
      </button>
      <button
        className="btn primary"
        onClick={() => {}}
      >
        삭제
      </button>
      <button
        type="submit"
        onClick={onSubmit}
        className="submit-btn btn secondary"
      >
        저장
      </button>
    </ModalFooter>
  </Modal>
);

CalModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  onToggle: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onDateChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  event: PropTypes.object.isRequired,
};

export default CalModal;
