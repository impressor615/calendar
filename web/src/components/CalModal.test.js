import React from 'react';
import moment from 'moment';

import CalModal from 'components/CalModal';
import { ModalHeader, ModalBody, ModalFooter } from 'components/Modal';


describe('<CalModal />', () => {
  const onToggle = jest.fn();
  const onChange = jest.fn();
  const onDateChange = jest.fn();
  const startDate = moment('2018-10-13');
  const endDate = moment('2018-10-13').add(1, 'hours');
  const props = {
    isOpen: false,
    onToggle,
    onChange,
    onDateChange,
    event: {
      title: 'title',
      start_date: startDate,
      end_date: endDate,
    },
  };

  it('should be rendered properly', () => {
    const wrapper = shallow(<CalModal {...props} />);
    expect(wrapper.find(ModalHeader)).toHaveLength(1);
    expect(wrapper.find(ModalBody)).toHaveLength(1);
    expect(wrapper.find(ModalFooter)).toHaveLength(1);
  });

  it('events should be fired properly', () => {
    const wrapper = shallow(<CalModal {...props} />);
    wrapper.find('input#title').simulate('change');
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('should be matched with the snapshot', () => {
    const wrapper = shallow(<CalModal {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});


