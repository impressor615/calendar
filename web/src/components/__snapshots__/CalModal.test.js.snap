// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalModal /> should be matched with the snapshot 1`] = `
<Modal
  className="cal-modal"
  isOpen={false}
>
  <ModalHeader
    onToggle={[MockFunction]}
  >
    일정 추가 및 변경
  </ModalHeader>
  <ModalBody>
    <input
      id="title"
      onChange={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
      placeholder="일정 제목을 입력해주세요!"
      type="text"
      value="title"
    />
    <DateRange
      endDate={"2018-10-12T16:00:00.000Z"}
      onChange={[MockFunction]}
      startDate={"2018-10-12T15:00:00.000Z"}
    />
  </ModalBody>
  <ModalFooter>
    <button
      className="btn primary"
      onClick={[MockFunction]}
    >
      취소
    </button>
    <button
      className="btn primary"
      onClick={[Function]}
    >
      삭제
    </button>
    <button
      className="submit-btn btn secondary"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
      type="submit"
    >
      저장
    </button>
  </ModalFooter>
</Modal>
`;
